targets/XPlatCppSdk/source/code/source/playfab/PlayFabAndroidHttpPlugin.cpp

```diff
@@ -0,0 +1,743 @@
+        { // LOCK httpRequestMutex
+            std::unique_lock<std::mutex> lock(httpRequestMutex);
+            requestTask->state = RequestTask::Pending;
+            pendingRequests.push_back(std::move(requestTask));
+            if(workerThread == nullptr)
+            {
+                threadRunning = true;
+                workerThread = std::make_unique<std::thread>(&PlayFabAndroidHttpPlugin::WorkerThreadEntry, this);
+            }
+        } // UNLOCK httpRequestMutex
+    }
+
+    size_t PlayFabAndroidHttpPlugin::Update()
+    {
+        if (PlayFabSettings::threadedCallbacks)
+        {
+            throw std::runtime_error("You should not call Update() when PlayFabSettings::threadedCallbacks == true");
+        }
+
+        std::shared_ptr<RequestTask> requestTask = nullptr;
+        { // LOCK httpRequestMutex
+            std::unique_lock<std::mutex> lock(httpRequestMutex);
+            if (pendingResults.empty())
+            {
+                return pendingRequests.size();
+            }
+
+            requestTask = std::move(this->pendingResults[0]);
+            this->pendingResults.pop_front();
+        } // UNLOCK httpRequestMutex
+
+        HandleResults(*requestTask);
+
+        // activeRequestCount can be altered by HandleResults, so we have to re-lock and return an updated value
+        { // LOCK httpRequestMutex
+            std::unique_lock<std::mutex> lock(httpRequestMutex);
+            return pendingRequests.size() + pendingRequests.size();
+        }
+    }
+
+    void PlayFabAndroidHttpPlugin::WorkerThreadEntry()
+    {
+        JavaVM* jvm;
+        if(!GetHelper().GetJavaVM(&jvm))
+        {
+            std::unique_lock<std::mutex> lock(httpRequestMutex);
+            threadRunning = false;
+            workerThread->detach();
+            workerThread = nullptr;
+            return ;
+        }
+        else
+        {
+            JNIEnv* jniEnv = nullptr;
+            jvm->AttachCurrentThread(&jniEnv, nullptr);
+
+            WorkerThread(static_cast<void*>(jniEnv));
+
+            jvm->DetachCurrentThread();
+        }
+    }
+
+    void PlayFabAndroidHttpPlugin::WorkerThread(void* jniEnv)
+    {
+        { // LOCK httpRequestMutex
+            std::unique_lock<std::mutex> lock(httpRequestMutex);
+            this->requestingTask = nullptr;
+        }
+        while (threadRunning) {
+            RequestTask::State state;
+            { // LOCK httpRequestMutex
+                std::unique_lock<std::mutex> lock(httpRequestMutex);
+                if (this->requestingTask == nullptr) {
+                    if (!pendingRequests.empty()) {
+                        this->requestingTask = this->pendingRequests[0];
+                        this->pendingRequests.pop_front();
+                        this->requestingTask->impl->SetJniEnv(static_cast<JNIEnv*>(jniEnv));
+                    } else {
+                        threadRunning = false;
+                        workerThread->detach();
```

Comment:
> workerThread->detach(); [](start = 23, length = 24)

Doesn't this mean you can fork any number of threads and exhaust the threadpool ?

Reply:
No, this is to reduce waiting time to close thread when the plug-in destroyed. This detach is release thread immediately, and if there MakePostRequest() is called during the closing thread, a new thread will create immediately without waiting to closing previouse thread.

---
In reply to: [274691406](https://github.com/PlayFab/SDKGenerator/pull/470#discussion_r274691406) [](ancestors = 274691406)